### ISSUE TITLE
marginal effects for interaction terms?

### ISSUE THREAD

**Issue by hauselin** [vincentarelbundock/marginaleffects#338]
Thanks for the amazing library! I'm wondering whether there's a way (or what's the best/correct way) to get the marginal effects for interaction terms? For example, I'm fitting the three-way interaction model `lm(mpg ~ hp * cyl * vs)`, and I'd like to know what the `hp:cyl` marginal effects are when `vs = 0` and `vs = 1`. 

In the code below, I show that when `vs = 0`, the estimate for `hp:cyl = 0.002953` (see model `m1`). When `vs = 1`, the estimate for `hp:cyl = 0.03468 ` (see model `m2`). Is it possible to use your library get these two marginal interaction effects without having to refit the models? As far as I can tell from the documentation, it doesn't look like it's possible, right? 

```r
# mean center predictors
mtcars$hp <- mtcars$hp - mean(mtcars$hp)
mtcars$cyl <- mtcars$cyl - mean(mtcars$cyl)

# 3-way interaction model
m1 <- lm(mpg ~ hp * cyl * vs, mtcars)
summary(m1)
Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 19.866261   1.804031  11.012 7.24e-11 ***
hp          -0.019580   0.039767  -0.492   0.6269    
cyl         -2.139013   1.153875  -1.854   0.0761 .  
vs          -2.278030   7.193554  -0.317   0.7542    
hp:cyl       0.002953   0.021023   0.140   0.8895   # hp:cyl effect for vs = 0
hp:vs       -0.017934   0.218246  -0.082   0.9352    
cyl:vs       1.322894   3.750573   0.353   0.7274    
hp:cyl:vs    0.031726   0.103082   0.308   0.7609  

# recode vs variable so 1 -> 0, and 0 -> 1
mtcars$vsR <- 1 - mtcars$vs
m2 <- lm(mpg ~ hp * cyl * vsR, mtcars)
summary(m2)
Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept) 17.58823    6.96367   2.526   0.0186 *
hp          -0.03751    0.21459  -0.175   0.8627  
cyl         -0.81612    3.56866  -0.229   0.8210  
vsR          2.27803    7.19355   0.317   0.7542  
hp:cyl       0.03468    0.10092   0.344   0.7341   # hp:cyl effect for vsR = 0 (i.e., vs = 1)
hp:vsR       0.01793    0.21825   0.082   0.9352  
cyl:vsR     -1.32289    3.75057  -0.353   0.7274  
hp:cyl:vsR  -0.03173    0.10308  -0.308   0.7609
```

**Comment by vincentarelbundock** [comment]
Thanks!

I'm not sure I understand what you need. Could you rephrase your query using the vocabulary that I use in the vignette to describe the quantities of interest:

https://vincentarelbundock.github.io/marginaleffects/#what

In particular, a "marginal effect" is defined as the slope of the outcome equation with respect to ONE regressor; it is a partial derivative.

Or maybe you are interested in "interactions between contrasts"? https://vincentarelbundock.github.io/marginaleffects/articles/mfx02_contrasts.html#interactions-between-contrasts

Or maybe you are interested in a Group-Average Marginal Effect, as defined here:

<https://vincentarelbundock.github.io/marginaleffects/articles/mfx03_mfx.html#group-average-marginal-effect-g-ame>

``` r
library(marginaleffects)
mod <- lm(mpg ~ hp * cyl * vs, mtcars)
mfx <- marginaleffects(mod)

summary(mfx, by = c("cyl", "vs"))
#> Average marginal effects 
#>    Term cyl vs   Effect Std. Error  z value Pr(>|z|)     2.5 %   97.5 %
#> 1    hp   6  0 -0.02013    0.04340 -0.46393 0.642700  -0.10519  0.06492
#> 2    hp   4  1 -0.11337    0.04753 -2.38529 0.017066  -0.20653 -0.02022
#> 3    hp   6  1 -0.04402    0.19618 -0.22436 0.822474  -0.42852  0.34049
#> 4    hp   8  0 -0.01423    0.01627 -0.87430 0.381955  -0.04612  0.01767
#> 5    hp   4  0 -0.02604    0.08389 -0.31039 0.756262  -0.19046  0.13838
#> 6   cyl   6  0 -2.18337    0.96819 -2.25509 0.024128  -4.08099 -0.28574
#> 7   cyl   4  1 -3.06634    3.40793 -0.89976 0.368246  -9.74576  3.61309
#> 8   cyl   6  1 -1.90633    1.21769 -1.56553 0.117459  -4.29296  0.48030
#> 9   cyl   8  0 -1.95439    2.26154 -0.86419 0.387486  -6.38693  2.47815
#> 10  cyl   4  0 -2.30344    0.93353 -2.46745 0.013608  -4.13313 -0.47375
#> 11   vs   6  0 -2.16734    3.89952 -0.55580 0.578350  -9.81025  5.47558
#> 12   vs   4  1  0.49506    3.39592  0.14578 0.884094  -6.16083  7.15095
#> 13   vs   6  1 -1.77527    2.03823 -0.87098 0.383764  -5.77013  2.21960
#> 14   vs   8  0  2.59387   37.50942  0.06915 0.944868 -70.92323 76.11098
#> 15   vs   4  0 -0.30842    3.00110 -0.10277 0.918147  -6.19047  5.57364
#> 
#> Model type:  lm 
#> Prediction type:  response
```

**Comment by hauselin** [comment]
Thanks! I'm interested in answering the question of what an **interaction effect** is at different levels of another variable. It's somewhat of a special case and my response below is rather long, so if my explanation still isn't clear, we can just close this issue. 

> In particular, a "marginal effect" is defined as the slope of the outcome equation with respect to ONE regressor; it is a partial derivative. 

That's right—ONE regressor is key here. Let me try to explain my question conceptually and with code again. Say I manually interact `hp` and `cyl` by multiplying them so I get ONE new regressor/column that captures their interaction, `hp_cyl`. 

In the code below, I've excluded irrelevant lines of output to keep things cleaner.

```r
# manually interact hp and cyl to create a new regressor hp_cyl that's their interaction
mtcars$hp_cyl <- mtcars$hp * mtcars$cyl  

# Q: what's the hp:cyl interaction effect when vs = 0
d0 <- mtcars[mtcars$vs == 0, ]
summary(lm(mpg ~ hp * cyl, d0))
summary(lm(mpg ~ hp + cyl + hp_cyl, d0))  # equivalent to above, now hp_cyl is ONE regressor
Coefficients:
             Estimate Std. Error t value Pr(>|t|)    
(Intercept) 19.866261   1.368487  14.517 7.84e-10 ***
hp_cyl       0.002953   0.015947   0.185   0.8558

# Q: what's the hp:cyl interaction effect when vs = 1
d1 <- mtcars[mtcars$vs == 1, ]
summary(lm(mpg ~ hp * cyl, d1))
summary(lm(mpg ~ hp + cyl + hp_cyl, d1))  # equivalent to above, now hp_cyl is ONE regressor
Coefficients:
            Estimate Std. Error t value Pr(>|t|)  
(Intercept) 17.58823    8.79298   2.000   0.0734 .
hp_cyl       0.03468    0.12743   0.272   0.7910  
```

So the question I'm trying to ask is _what's the two-way `hp:cyl` interaction effect at different levels of `vs`_? In fact, if I were to use the interaction regressor/column `hp_cyl` I manually created by multiplying `hp` and `cyl` columns, I can actually use `marginaleffects()` to answer this question, without having to fit two separate models (see above, one for `vs = 0` and the other for `vs = 1`).

```r
# proper three-way interaction model
mod1 <- lm(mpg ~ hp * cyl * vs, mtcars)
summary(mod1)

# same three-way interaction model but using my manually coded hp_cyl column/regressor
mod2 <- lm(mpg ~ (hp + cyl + hp_cyl) * vs, mtcars)
summary(mod2)  # same results as summary(mod1)

# Q: what's the effect of hp_cyl (i.e., hp:cyl in mod1) for vs = 0 and vs = 1?
marginaleffects(mod2, newdata = datagrid(vs = c(0, 1)))  # showing only the two relevant rows
  rowid     type   term         dydx
5     1 response hp_cyl  0.002952685  # vs = 0
6     2 response hp_cyl  0.034678728  # vs = 1
```

In `mod2`, because I manually created the column `hp_cyl` that captures the two-way interaction between `hp` and `cyl`, it's a single regressor/column that I can then interact with `vs`, and I can then call `marginaleffects(mod2, newdata = datagrid(vs = c(0, 1)))` to get the "marginal effect" of `hp_cyl` for `vs = 0` and `vs = 1`. The `dydx` values are in fact identical to when I fit separate models to subsets of the data. 

```r
summary(lm(mpg ~ hp * cyl, mtcars[mtcars$vs == 0, ]))
             Estimate Std. Error t value Pr(>|t|)    
hp:cyl       0.002953   0.015947   0.185   0.8558    

summary(lm(mpg ~ hp * cyl, mtcars[mtcars$vs == 1, ]))
             Estimate Std. Error t value Pr(>|t|)    
hp_cyl       0.03468    0.12743   0.272   0.7910   

mod2 <- lm(mpg ~ (hp + cyl + hp_cyl) * vs, mtcars)
marginaleffects(mod2, newdata = datagrid(vs = c(0, 1)))  # showing only the two relevant rows
  rowid     type   term         dydx
5     1 response hp_cyl  0.002952685  # vs = 0
6     2 response hp_cyl  0.034678728  # vs = 1
# but can't get these results (hp:cyl for vs = 0 and vs = 1) if the model was lm(mpg ~ hp * cyl * vs), unless I manually recode the vs values (as in the code in my initial question)
```

Again, if all this still doesn't make sense, we can close this issue, since I think it's quite a special case. Thanks again for the incredible libraries (this and `modelsummary`)!

**Comment by hauselin** [comment]
Forgot to mention, I also read this page (https://vincentarelbundock.github.io/marginaleffects/reference/comparisons.html) and think `comparisons(model, variables = list(hp = c(...), cyl = c(...), vs = c(...)))` might answer my question, but I haven't tested it yet (too many neat features built into library!).

**Comment by vincentarelbundock** [comment]
Thanks for the details!

Frankly, I'm still a bit confused. Two issues:

1) You have to be very careful when creating new hard-coded variables like this. If you multiply `cyl` and `hp` ahead to time to create `hp_cyl`, then `marginaleffects

**Comment by hauselin** [comment]
Thanks for looking into this!
